### ISSUE TITLE
Add paging support to tools metric

### ISSUE THREAD

**Issue by sourabhtk37** [distributed-system-analysis/pbench#1200]
While viewing tools data (eg: proc-interrupts-per-cpu) the page doesn't load up due to large amount of data that has to be loaded by browser in some cases(eg: systems with large number of CPUs).

Idea is to send the data in pages to avoid crashing the browser.

**Comment by atheurer** [comment]
We could update proc-interrupts-postprocess to generate 1 html page per CPU, and/or 1 html page per IRQ number.  There will be hundreds of html pages, however.

**Comment by k-rister** [comment]
Alternatively, we could just tweak the page such that each graph is replaced with a header.  Clicking a header would remove an existing graph, if present, and show the requested graph.  That way a single page would still be generated (single pane) but it would searchable (ie. "cpu 6" on a per cpu page) and it would significantly minimize the page concurrency by only loading one graph at a time into the DOM.  On the flip side, this idea (and the one graph per page idea) would further limit the ability to correlate between graphs than what is currently already there without using more windows and/or tabs.

**Comment by sourabhtk37** [comment]
@atheurer This will apply to many metric data that shows lot of graph details eg: proc-interrupts, sar etc

@k-rister Yeah, data isolation would be a problem. What I can think of is hover for "Click to load" on each graph of proc-interrupts, sar etc On clicking, data is loaded.

**Comment by k-rister** [comment]
@sourabhtk37 Yeah, we could probably just make my idea a little bit more flexible with each graph being loaded on demand (ie. on a click) but only hidden if the user requests with another action as opposed to automatically.  So if a particular page is causing problems, the user could reload it and then only show one graph at a time.

**Comment by portante** [comment]
@k-rister, @gurbirkalsi, are there any specific plans to address this problem?

**Comment by k-rister** [comment]
@portante I have the basic idea formulated in my head but I haven't spent any time on it, yet.

My current thinking is to change the generic library GenData.pm to create buttons on the page rather than charts.  If you click the button it will then render the chart.  If you click the button again it will destroy the chart.  This would, by default, make page loads really fast since nothing would be generated automatically.  It would allow the user to only see the graphs on a page that they are interested in rather than all of them.  If they load too many graphs and it blows up they simply reload the page and reduce what they display on the next try.

That said, there are several caveats and/or downsides to this:

- People that are loading a page and used to seeing all of the charts would no longer get that behavior, but doing this in reverse (where you hide the ones you don't want to see) will likely not help to resolve the problem.  We have to preemptively avoid this problem rather than react to it in order to avoid it.
- This would help alleviate problems where the issue is the total amount of data loaded and/or rendered across several graphs.  It would not do anything to solve the problem if a single graph is overwhelming the browser.

The real solution we are working on is in the CDM work with ES where we will be interactively limiting the datasets and dataset size that is pulled to the browser for rendering.  That is something that I'm working toward with @atheurer and we are inching closer but we aren't there yet.  In theory this should give the ideal solution that doesn't have downsides for certain uses cases.

My preference would be to focus on the real solution and get there ASAP, but if we feel that a temporary fix is truly necessary then I can work on that too.
